Library/Homebrew/cmd/list.sh

```diff
@@ -0,0 +1,92 @@
+#:  * `list`
+#:
+#: List all installed formulae and casks.
+#: If <formula> is provided, summarise the paths within its current keg.
+#: If <cask> is provided, list its artifacts.
+
+# HOMEBREW_CELLAR, HOMEBREW_PREFIX are set by brew.sh
+# shellcheck disable=SC2154
+
+homebrew-list() {
+  case "$1" in
+    # check we actually have list and not e.g. listsomething
+    list) ;;
+    list*) return 1 ;;
+    *) ;;
+  esac
+
+  local ls_args=()
+  local formula=""
+  local cask=""
+
+  while [[ "$#" -gt 0 ]]
+  do
+    case "$1" in
+      # check for flags passed to ls
+      -1 | -l | -r | -t)
+        ls_args+=("$1")
+        shift
+        ;;
+      --formula | --formulae)
+        formula=1
+        shift
+        ;;
+      --cask | --casks)
+        cask=1
+        shift
+        ;;
+      # reject all other flags
+      *) return 1 ;;
+    esac
+  done
+
+  local tty
+  if [[ -t 1 ]]
+  then
+    tty=1
+  fi
+
+  local error_string="LS_ERRORED"
+  if [[ -z "${cask}" && -d "${HOMEBREW_CELLAR}" ]]
+  then
+    if [[ -n "${tty}" && -z "${formula}" ]]
+    then
+      ohai "Formulae"
+    fi
+
+    local formula_output
+    formula_output="$(ls "${ls_args[@]}" "${HOMEBREW_CELLAR}" || echo "${error_string}")"
+    if [[ "${formula_output}" == "${error_string}" ]]
+    then
+      return 1
+    elif [[ -n "${formula_output}" ]]
+    then
+      echo "${formula_output}"
```

Comment:
I think we should capture the output of both `ls` calls first before outputting anything, else we could end up in the awkward situation where `ls ${HOMEBREW_CELLAR}` succeeds but `ls ${HOMEBREW_CASKROOM}` doesn't, which means we've printed the formulae list already by the time we falls back to `list.rb`.

Reply:
@carlocab I've changed these `return`s to `exit`. They are shelling out to identical commands to in Ruby so I agree that `return` and then running the same command doesn't make sense, particularly given we're going to print errors here (and I don't think it makes sense to eat them).